Keep only important virtual methods in drivers

MikeAlhayek opened the pull request:
<!--- Please make sure that you're familiar with our contribution guidelines before submitting a pull request: https://docs.orchardcore.net/en/latest/guides/contributing/. -->

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.AdminMenu/AdminNodes/PlaceholderAdminNodeDriver.cs:
```diff
@@ -18,44 +17,49 @@ public PlaceholderAdminNodeDriver(IAdminMenuPermissionService adminMenuPermissio
         }
 
-        public override IDisplayResult Display(PlaceholderAdminNode treeNode)
+        public override Task<IDisplayResult> DisplayAsync(PlaceholderAdminNode treeNode, BuildDisplayContext context)
```
@Piedone we had the same argument during the last meeting. @sebastienros argument was following the Microsoft standard. Microsoft never implement sync and async version of the same method. Which kinda makes sense. As you see I exposed CombineAsync method to reduce the code written to leave only one method which will be the best was for teaching someone and also for performance.

MikeAlhayek replied:
Oh, and if we want to do that, we'll have to introduce a breaking change to add the sync versions that accept the context.

Piedone replied:
It's frequent in .NET to have both though, like all the file I/O (including methods introduced in later .NET versions) or `IDistributedCache`. But that aside, my point is that if the consumers of our API will frequently write only sync code, which is apparently the case, then we should ideally make that easier (and/or even consider `ValueTask`, but I don't think we can have enough data to rule for or against that). `CombinedAsync` is indeed good, but there's still a lot of awkward `Task.FromResult` still.

But I see the issue with the call chain. There must be a way out of this somehow, this looks simple... The only one I can think of though is adding a condition to each override like this:

```c#
public virtual async Task<IDisplayResult> DisplayAsync(TModel model, TDisplayContext context)
{
    return Display(model, context) ?? await DisplayAsync(model, context.Updater);
}

public virtual IDisplayResult Display(TModel model, TDisplayContext context)
{
    return Display(model, context.Updater);
}

...
```

But this is ugly and adds branching to something called very frequently.

Or, we accept that this won't be great in one go, merge this PR, and introduce an `IDisplayResult Display(TModel model, TDisplayContext context)` only in 3.0.

Did you figure out something better today? Hence https://github.com/OrchardCMS/OrchardCore/pull/16508#issuecomment-2263655591 @sebastienros?

@sarahelsaig as Lombiq's Chief Orchard Core Upgrade Officer, could you please take a look and provide feedback?

sarahelsaig replied:
I approve of marking all the sync versions obsolete. It keeps the API simpler and I still remember that having both overridables in the base class was confusing back when I started working with OC.

I don't think returning `Task.FromResult()` is not a huge ask. Could we add an `abstract class SyncDisplayDriver<TModel, TDisplayContext, TEditorContext, TUpdateContext> : DisplayDriver<TModel, TDisplayContext, TEditorContext, TUpdateContext>` type to OrchardCore.DisplayManagement? That would be a nice way to bridge the gap: when you are sure you are only doing sync stuff in a driver you can inherit from it instead of the base `DisplayDriver`. If you need both sync and async methods in the same driver, then you can put up with the extra `Task.FromResult()`.

Piedone replied:
Hmm, having a fully sync base driver is an interesting idea! Though note that `DisplayDriver<TModel, TDisplayContext, TEditorContext, TUpdateContext>` is only one of the base classes, and while lower-level drivers directly inherit from it, we also have `ContentPartDisplayDriver<TPart>` and `ContentFieldDisplayDriver<TField>`. Those have similar issues but seem to be more consistent with offering sync overloads.

sarahelsaig replied:
Yeah, in the linked PR I've also included sync versions is both of the classes you mentioned, and altered a few classes to use them.

Piedone replied:
Didn't you forget to push? Because I don't see part or field drivers there.

MikeAlhayek replied:
@Piedone @sebastienros @sarahelsaig please check out the updated. Now, every important async method has a corresponding sync method. When possible, I marked old methods as obsolete. However, it was necessary to remove some methods to accomplish this.

I did the same change to the `SectionDisplayDriver` class. Also, simplified the site settings by introducing a new base class `SiteDisplayDriver<SomeSettings>`. 

Please review the changes and provide feedback.

MikeAlhayek replied:
@Piedone  I think @sarahelsaig was talking about another draft PR https://github.com/OrchardCMS/OrchardCore/pull/16514.

Personally, I like to use a single driver for this so we don't add any more confusion. I think having one sync and one async version is good approach for this.

Piedone replied:
Yep, I've check that PR. Anyway, yeah, all in all single base drivers would be better.

MikeAlhayek replied:
Good I think we are all on the same page. Can you please give this PR a review hoping to merge it soon.

Piedone replied:
Yeah, on it, just it's quite a lot.

MikeAlhayek replied:
Yes it is alot. Obviously everything is important, but the key changes to review is `SectionDisplayDriver` and `DisplayDriver`

MikeAlhayek replied:
I am resolving this since we all agree on the approach.

sebastienros commented:
ok with extra sync ones

sebastienros commented:
> It's frequent in .NET to have both though, like all the file I/O

When something provides IO implementations it makes sense to have "different" implementation, because the caller can choose. Here we are asking for users to provide overrides, not for different implementations, the async will call the sync one. This is different.

Piedone commented on src/OrchardCore/OrchardCore.DisplayManagement/Entities/SectionDisplayDriver.cs:
```diff
@@ -34,107 +38,111 @@ public override Task<IDisplayResult> DisplayAsync(TModel model, BuildDisplayCont
+#pragma warning restore CS0618 // Type or member is obsolete
         }
 
+        [Obsolete("This method will be removed. Instead use UpdateAsync(TModel model, TSection section, UpdateEditorContext context)")]
```
Since there's no such method, I believe you meant `UpdateAsync(TModel model, UpdateEditorContext context)`. Also for the other `Obsolete` below.

MikeAlhayek replied:
The referenced method exists on line 113.

Piedone commented on src/docs/releases/2.0.0.md:
```diff
@@ -269,7 +269,52 @@ Here are the updated signatures:
    `Task<IDisplayResult> UpdateAsync(TSection section, UpdateEditorContext context)`
 
-These adjustments ensure compatibility and adherence to the latest conventions within the `SectionDisplayDriver` class.
+These adjustments ensure compatibility and adherence to the latest conventions within the `DisplayDriver` class.
```
The "To" methods mentioned under points 1 and 2 above don't exist anymore.

MikeAlhayek replied:
In the `SectionDisplayDriver` we want people to use `UpdateAsync(TModel model, TSection section, UpdateEditorContext context)`.

In the `DisplayDriver` we want people to use `UpdateAsync(TModel model, UpdateEditorContext context)`

Both of these methods exists,

Piedone replied:
Under point 1 the To method is `Task<IDisplayResult> UpdateAsync(TModel model, TSection section, IUpdateModel updater, UpdateEditorContext context)` in `SectionDisplayDriver`. There's no such method.

Point 2 refers to `Task<IDisplayResult> UpdateAsync(TSection section, IUpdateModel updater, UpdateEditorContext context)` in `SectionDisplayDriver`. There's no such method either.

MikeAlhayek replied:
I think I am not understanding here.

In the `SectionDisplayDriver`, the update obsolete message states the following:

> This method is obsolete and will be removed in version 3. Instead, use the UpdateAsync(TModel model, TSection section, UpdateEditorContext context)

This method exists here:

[        public virtual Task<IDisplayResult> UpdateAsync(TModel model, TSection section, UpdateEditorContext context)](https://github.com/OrchardCMS/OrchardCore/blob/53869c4d516c997ae334fbdf2e84421298c0698f/src/OrchardCore/OrchardCore.DisplayManagement/Entities/SectionDisplayDriver.cs#L111C27-L111C117)

In the 2.0.0 release notes we list the methods that were removed like the one you mentioned in your comment:

```
 and **removed** the following methods:

- `UpdateAsync(TModel model, TSection section, IUpdateModel updater, UpdateEditorContext context)`
- `UpdateAsync(TSection section, IUpdateModel updater, UpdateEditorContext context)`
- `UpdateAsync(TSection section, IUpdateModel updater, string groupId)`
- `UpdateAsync(TSection section, UpdateEditorContext context)`
- `EditAsync(TSection section, BuildEditorContext context)`
- `DisplayAsync(TSection section, BuildDisplayContext context)`
```

What am I missing here?

Piedone replied:
There's this section in the release notes I was referring to in my opening comment:

![image](https://github.com/user-attachments/assets/00096bac-a4f7-4c56-ad79-17f14bcb1a12)

Point 1 and 2 are now inaccurate.

MikeAlhayek replied:
ahhh. Got it. I fixed this.

Piedone commented on src/docs/releases/2.0.0.md:
```diff
@@ -269,7 +269,52 @@ Here are the updated signatures:
+ 
+
+ !!! note
+    With the new `SiteDisplayDriver` base class for custom settings, you no longer need explicit checks like `context.GroupId.Equals(GroupId, StringComparison.OrdinalIgnoreCase)` as the base class handles this for you.
```
`SettingsGroupId` needs to be overridden for this, which is missing from the below samples.

MikeAlhayek replied:
both implementation in the 2.0.0 release of `SiteDisplayDriver` are stripped versions. Both state the following

```
    //
    // For  example simplicity, other methods are not visible.
    //
```

Piedone reviewed (changes requested):
I did some testing, but only a cursory review apart from the base classes and docs.

MikeAlhayek commented on src/OrchardCore/OrchardCore.DisplayManagement/Entities/SectionDisplayDriver.cs:
```diff
@@ -18,14 +19,17 @@ public abstract class SectionDisplayDriver<TModel, TSection> : DisplayDriver<TMo
         where TModel : class, IEntity
     {
+        [Obsolete("Instead use SectionName")]
+        protected virtual string PropertyName => typeof(TSection).Name;
```
I changed it back to `PropertyName` to avoid having to debate which is a better name. But I think SectionName would make more sense here since we are talking about section not property.

MikeAlhayek commented on src/OrchardCore/OrchardCore.DisplayManagement/Entities/SectionDisplayDriver.cs:
```diff
@@ -18,14 +19,17 @@ public abstract class SectionDisplayDriver<TModel, TSection> : DisplayDriver<TMo
+
         /// <summary>
-        /// Gets the property name that the <typeparamref name="TSection"/> is stored on the <typeparamref name="TModel"/>.
+        /// Gets the section name that the <typeparamref name="TSection"/> is stored on the <typeparamref name="TModel"/>.
```
I reverted this change and update the comment.

MikeAlhayek commented:
@Piedone I think I addressed all of your comments.

I also just made a change to reduce the amount of methods we are able to override in the `SectionDisplayDriver`. I moved all the code that was in `SectionDisplayDriver` into a new class called `SectionDisplayDriverBase`. Then inhireted the `SectionDisplayDriver` from the new `SectionDisplayDriverBase` and sealed the 3 methods that one should not be able to override by design.

The `SectionDisplayDriver` look like this now

```
public abstract class SectionDisplayDriver<TModel, TSection> : SectionDisplayDriverBase<TModel, TSection>
    where TSection : new()
    where TModel : class, IEntity
{
    public sealed override Task<IDisplayResult> DisplayAsync(TModel model, BuildDisplayContext context)
        => base.DisplayAsync(model, context);

    public sealed override Task<IDisplayResult> EditAsync(TModel model, BuildEditorContext context)
        => base.EditAsync(model, context);

    public sealed override Task<IDisplayResult> UpdateAsync(TModel model, UpdateEditorContext context)
        => base.UpdateAsync(model, context);
}
```

Piedone commented:
I don't fully understand. Why shouldn't those three method be possible to override?

MikeAlhayek commented:
> I don't fully understand. Why shouldn't those three method be possible to override?

If you are overriding the `SectionDisplayDriver` you should be working with sections and not the model directly. If you need the model, then you should be inheriting from `DisplayDriver` not `SectionDisplayDriver`. This reduce the methods one can override in the `SectionDisplayDriver` and reduce the confusions. + since these are sealed, we should be slightly better performance.

Piedone reviewed (approved):
OK then! LGTM but I'd like to see what the others think.

MikeAlhayek commented:
Sounds good. We'll give @sebastienros and @sarahelsaig few days to provide any feedback and we'll merge it Monday if we don't hear back.
